### PR TITLE
Fix a false positive in IDE0059 in presence of try/finally

### LINKED ISSUE #39344: IDE0059 "Remove redundant assignment" can break code logic
**Version Used**: Version 16.4.0 Preview 3.0 [29416.158.master]

**Steps to Reproduce**:

```c#
void M(int i)
{
    bool b = false;
    try
    {
        if (i == 0)
        {
            b = true; // IDE0059 Unnecessary assignment of a value
        }
    }
    finally
    {
        if (!b)
        {
            Console.WriteLine("Exception");
        }
    }
}
```

**Expected Behavior**:

Does not remove assignment to `b`.

**Actual Behavior**:

Removes assignment to `b`.

### LINKED ISSUE #39755: Regression: false positive IDE0059 in 16.4
**Version Used**: 
Visual Studio 16.3.8, 16.4 Preview 4.0

**Steps to Reproduce**:
```csharp
        void M()
        {
            IntPtr a = (IntPtr)1;
            try
            {
                var b = a;

                if(Some(a))
                    a = IntPtr.Zero; // here
            }
            finally
            {
                if (a != IntPtr.Zero)
                {

                }
            }
        }

        bool Some(IntPtr a) => true;
```

**Expected Behavior**:
No IDE0059 reported, in 16.3.

**Actual Behavior**:
False positive IDE0059 redundant assignment reported, in 16.4.

Only regresses when the consumptions of `b` and `if(Some(a))` are all present.

### PR DESCRIPTION
**Pull request by mavasani**
We were aggressively trimming down tracked locals when encountering a branch that leaves a try/finally region. The fix makes us more conservative for such cases.

Fixes #39344
Fixes #39755

Verified that both the added unit tests fail before the product fix.

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs [hunk 1]
```diff
@@ -7239,5 +7239,67 @@ string M(object obj)
     }
 }", options: PreferDiscard, parseOptions: new CSharpParseOptions(LanguageVersion.CSharp8));
         }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnusedValues)]
+        [WorkItem(39344, "https://github.com/dotnet/roslyn/issues/39344")]
+        public async Task AssignmentInTry_UsedInFinally_NoDiagnostic()
+        {
+            await TestDiagnosticMissingAsync(
+@"using System;
```
Comment (ryzngard) [inline src/EditorFeatures/CSharpTest/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs]: Do we expect to report if it's not used in the finally? 

```c# 
void M(int i)
{
    bool b = false;
    try 
    { 
        if (i == 0) b = true;
    }
    finally 
    {
    }
}
```

Comment (mavasani) [reply src/EditorFeatures/CSharpTest/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs]: Yes, that works prior to the fix - both assignments are highlighted as redundant. I can still add a unit test.

**Comment by mavasani** [comment]
@ryzngard Let me know if you have any other feedback. Thanks!

**Comment by mavasani** [comment]
@dotnet/roslyn-infrastructure - do we know why the `license/cla` approval is stuck on PRs?

**Comment by ryzngard** [comment]
Potentially related to https://status.dev.azure.com/_event/162555283

**Pull request merged**